### PR TITLE
istioctl: we should warn when namespace is not labeled ambient

### PR DESCRIPTION
**Pull request by MorrisLaw**
**Please provide a description of this PR:**
Resolves: #50396

We are not checking if a namespace is labeled ambient before we enroll a waypoint with a namespace. This PR adds that check that returns early before we apply the waypoint if `--enroll-namespace` flag is used and the namespace does not have that `ambient` label.

This is what the UX will look like now:

```
azureuser@dev-2024:~/istio$ istioctl x waypoint apply --enroll-namespace                                            
Warning: namespace is not enrolled in ambient. Consider running `kubectl label namespace default istio.io/dataplane-mode=ambient`     

azureuser@dev-2024:~/istio$ istioctl x waypoint apply --enroll-namespace                                            
Warning: namespace is not enrolled in ambient. Consider running `kubectl label namespace default istio.io/dataplane-mode=ambient`     

azureuser@dev-2024:~/istio$ istioctl x waypoint apply --enroll-namespace                                         
Warning: namespace is not enrolled in ambient. Consider running `kubectl label namespace default istio.io/dataplane-mode=ambient`     

azureuser@dev-2024:~/istio$ kubectl label namespace default istio.io/dataplane-mode=ambient                                           
namespace/default labeled

azureuser@dev-2024:~/istio$ istioctl x waypoint apply --enroll-namespace                        
waypoint default/default-namespace applied
namespace default annotated with waypoint default-namespace  
```

### CODE REVIEW

**Comment by ilrudie** [comment]
I don't think we want to return early/fail if the namespace is not labeled with dataplane-mode=ambient. There are some niche, but valid, uses to allowing it to proceed with warning IMO.

File: istioctl/pkg/waypoint/waypoint.go [hunk 1]
```diff
@@ -414,3 +428,19 @@ func annotateNamespaceWithWaypoint(kubeClient kube.CLIClient, ns string) error {
...
+	} else if err != nil {
+		return false, fmt.Errorf("failed to get namespace %s: %v", ns, err)
+	}
+	if nsObj.Labels == nil {
+		return false, nil
+	}
+	if nsObj.Labels[constants.DataplaneMode] == constants.DataplaneModeAmbient {
+		return true, nil
+	}
+	return false, nil
```
Comment (ilrudie) [inline istioctl/pkg/waypoint/waypoint.go:445]: possible change: not sure this actually reads any more clearly
```suggestion
	return nsObj.Labels[constants.DataplaneMode] == constants.DataplaneModeAmbient, nil
```

**Review by ilrudie — COMMENTED** [review]
aside from the question of whether or not to fail or simply warn this seems good to me

File: istioctl/pkg/waypoint/waypoint.go (hunk 1 above)
Comment (MorrisLaw) [reply istioctl/pkg/waypoint/waypoint.go:445]: This is cleaner tbh. Will update

File: istioctl/pkg/waypoint/waypoint.go [hunk 2]
```diff
@@ -140,6 +140,19 @@ func Cmd(ctx cli.Context) *cobra.Command {
...
+			ns := ctx.NamespaceOrDefault(ctx.Namespace())
+			// If a user decides to enroll their namespace with a waypoint, verify that they have labeled their namespace as ambient.
+			if enrollNamespace {
+				namespaceIsLabeledAmbient, err := namespaceIsLabeledAmbient(kubeClient, ns)
+				if err != nil {
+					return fmt.Errorf("failed to check if namespace is labeled ambient: %v", err)
+				}
+				if !namespaceIsLabeledAmbient {
+					fmt.Fprintln(cmd.OutOrStdout(), "Warning: namespace is not enrolled in ambient. Consider running\t"+
+						"`"+"kubectl label namespace default istio.io/dataplane-mode=ambient"+"`")
```
Comment (howardjohn) [inline istioctl/pkg/waypoint/waypoint.go:152]:
```suggestion
						"`"+"kubectl label namespace %s istio.io/dataplane-mode=ambient"+"`")
```

(and plug in `ns`)

**Comment by MorrisLaw** [comment]
> I don't think we want to return early/fail if the namespace is not labeled with dataplane-mode=ambient. There are some niche, but valid, uses to allowing it to proceed with warning IMO.

I suppose letting the user do what they requested to do is fine. As long as they know that they didn't label the namespace; we're also giving them the command to do so if that makes sense for them. As for those niche use cases, then they just won't label. This makes sense to me, I'll update it.

**Comment by ilrudie** [comment]
@MorrisLaw I think the most straight forward use would be if the users are handling ambient redirection per WL rather than at the namespace level so there are pods with ambient redirection enabled but the namespace label isn't present. IMO we don't need to check a case like that here though

**Comment by hzxuzhonghu** [comment]
Trying to understand, waypoint does not support proxying an access to a service/workload, which isn't in charge by ambient.
So both client and server should be in ambient charge, not a single side. right?

File: releasenotes/notes/50395.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: feature
+area: istioctl
+issue:
+  - 50396
+releaseNotes:
+- |
+  **Added** Warn user in standard out when `x waypoint apply --enroll-namespace` is called and namespace is not labeled ambient.
```
Comment (hanxiaop) [inline releasenotes/notes/50395.yaml:8]:
```suggestion
  **Added** Warn user in standard out when `istioctl experimental waypoint apply --enroll-namespace` is called and namespace is not labeled ambient.
```

**Comment by ilrudie** [comment]
> Trying to understand, waypoint does not support proxying an access to a service/workload, which isn't in charge by ambient. So both client and server should be in ambient charge, not a single side. right?

In the most common case that's correct. Both client and server should be enrolled in ambient (captured). I think anything with client enrollment is out of scope for this PR though since client-side is likely not controlled by the waypoint owner and could be very diverse depending on environment.

We had also talked about letting a waypoint have non-hbone listeners so that uncaptured clients could still target the waypoint which changed the calculus for clients as well. I am not sure the state of this use case right now though.

**Comment by howardjohn** [comment]
You can have a waypoint for a non-captured workload currently. Not
recommended, though

**Review by linsun — COMMENTED** [review]
LGTM, thanks Mike!

**Comment by MorrisLaw** [comment]
> LGTM, thanks Mike!

I'm the other Morris from Microsoft, Jeremy (;

**Pull request merged**